src/lifo.zig

```diff
@@ -0,0 +1,216 @@
+const std = @import("std");
+const assert = std.debug.assert;
+
+const constants = @import("./constants.zig");
+
+/// An intrusive last in/first out (LIFO, stack) linked list.
+/// The element type T must have a field called "next" of type ?*T.
+pub fn LIFOType(comptime T: type) type {
+    return struct {
+        const LIFO = @This();
+
+        head: ?*T = null,
+
+        count: u64 = 0,
+        count_max: u64,
+
+        // This should only be null if you're sure we'll never want to monitor `count`.
+        name: ?[]const u8,
+
+        // If the number of elements is large, the constants.verify check in push() can be too
+        // expensive. Allow the user to gate it. Could also be a comptime param?
+        verify_push: bool,
+
+        pub fn init(count_max: u64, name: ?[]const u8, verify_push: bool) LIFO {
+            return LIFO{
+                .name = name,
+                .count_max = count_max,
+                .verify_push = verify_push,
+            };
+        }
+
+        /// Pushes a new node to the first position of the LIFO.
+        pub fn push(self: *LIFO, node: *T) void {
+            if (constants.verify and self.verify_push) assert(!self.contains(node));
+
+            assert((self.count == 0) == (self.head == null));
+            assert(node.next == null);
+            assert(self.count < self.count_max);
+
+            // Insert the new element at the head.
+            node.next = self.head;
+            self.head = node;
+            self.count += 1;
+        }
+
+        /// Returns the first element of the stack, and removes it.
+        pub fn pop(self: *LIFO) ?*T {
+            assert((self.count == 0) == (self.head == null));
+
+            const node = self.head orelse return null;
+            self.head = node.next;
+            node.next = null;
+            self.count -= 1;
+            return node;
+        }
+
+        /// Returns the first element of the stack, but does not remove it.
+        pub fn peek(self: LIFO) ?*T {
+            return self.head;
+        }
+
+        /// Checks if the LIFO is empty.
+        pub fn empty(self: LIFO) bool {
+            assert((self.count == 0) == (self.head == null));
+            return self.head == null;
+        }
+
+        /// Returns whether the linked list contains the given *exact element* (pointer comparison).
+        fn contains(self: *const LIFO, needle: *const T) bool {
+            var iterator = self.head;
+            var count: u64 = 1;
+            while (iterator) |node| : (iterator = node.next) {
+                if (node == needle) return true;
+                count += 1;
+                assert(count <= self.count_max);
+            }
+            return false;
```

Comment:
Let's go for a for loop here to make it _syntactically_ obvious that the loop is bounded: 

```suggestion
            assert(self.count <= self.count_max);
            var next = self.head;
            for (0..self.count + 1) |_| {
                const node = next orelse return false;
                if (node == needle) return true;
                next = node.next;
            } else unreachable;
```

As a bonus, this also verifies that `self.count` is correct!